src/networking/payloads/introductionrequestpayload.rs

```diff
@@ -32,125 +34,70 @@ struct IntroductionRequestPayload {
-
-    let conntype = self.connection_type.encode();
-    res
-      .add_raw(
-        vec![
-          destination_address[0],
-          destination_address[1],
-          destination_address[2],
-          destination_address[3],
-        ],
-        4,
-      )
-      .add_u16(destination_port)
-      .add_raw(
-        vec![
-          source_lan_address[0],
-          source_lan_address[1],
-          source_lan_address[2],
-          source_lan_address[3],
-        ],
-        4,
-      )
-      .add_u16(source_lan_port)
-      .add_raw(
-        vec![
-          source_wan_address[0],
-          source_wan_address[1],
-          source_wan_address[2],
-          source_wan_address[3],
-        ],
-        4,
-      )
-      .add_u16(source_wan_port)
-      .add_bits(Bits::from_bools((
-        conntype.0,
-        conntype.1,
-        false,
-        false,
-        false,
-        false,
-        false,
-        self.advice,
-      )))
-      .add_u16(self.identifier)
-      .add_raw_remaining(self.extra_bytes.clone());
-    res
+impl Ipv8Payload for IntroductionRequestPayload{
+  /// this function is necessary for any payload which has a rawend final field. It is called to set this field. (like a setter.)
+  fn set_rawend(&mut self, bytes:RawEnd){
+    self.extra_bytes = bytes;
   }
+}
 
-  fn unpack(packet: Packet) -> Self {
-    let mut packetiter = packet.iter();
-
-    let destination_ipaddress = packetiter.next_raw(4).unwrap();
-    let destination_port = packetiter.next_u16().unwrap();
-
-    let source_lan_address = packetiter.next_raw(4).unwrap();
-    let source_lan_port = packetiter.next_u16().unwrap();
-
-    let source_wan_address = packetiter.next_raw(4).unwrap();
-    let source_wan_port = packetiter.next_u16().unwrap();
-
-    // flags are cc00000a where cc is two bits signifying the connection type and a signifies the advice
-    let flags = packetiter.next_bits().unwrap();
-
-    let identifier = packetiter.next_u16().unwrap();
-
-    let extra_bytes = packetiter.next_raw_remaining().unwrap();
+/// makes the IntroductionRequestPayload serializable.
+/// This is less than trivial as there is no 1:1 mapping between the serialized data and the payload struct.
+/// Some struct fields are combined into one byte to form the serialized data.
+impl Serialize for IntroductionRequestPayload {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where S: Serializer{
+        let conntype = self.connection_type.encode();
+
+        let mut state = serializer.serialize_struct("IntroductionRequestPayload", 6)?;
+        state.serialize_field("destination_address", &self.destination_address)?;
+        state.serialize_field("source_lan_address", &self.source_lan_address)?;
+        state.serialize_field("source_wan_address", &self.source_wan_address)?;
+        state.serialize_field("advice", &Bits::from_bools((conntype.0, conntype.1, false, false, false, false, false, self.advice)))?;
```

Comment:
A bunch of magic `false` statements. I have no clue what these mean. Do you? If you do, please consider either putting these into named variables or add a line comment that describes what these mean.

Reply:
Will add a comment.